lib/algo/DMD.cpp

```diff
@@ -343,11 +357,112 @@ DMD::constructDMD(const Matrix* f_snapshots,
         d_S_inv->item(i, i) = 1 / d_factorizer->S[static_cast<unsigned>(i)];
     }
 
+    Matrix* Q = NULL;
+
+    // If W0 is not null, we need to ensure it is in the basis of W.
+    if (W0 != NULL)
+    {
+        CAROM_VERIFY(W0->numRows() == d_basis->numRows());
+        CAROM_VERIFY(linearity_tol >= 0.0);
+        std::vector<int> lin_independent_cols_W;
+
+        // Copy W0 and orthogonalize.
+        Matrix* d_basis_init = new Matrix(f_snapshots->numRows(), W0->numColumns(), true);
+        for (int i = 0; i < d_basis_init->numRows(); i++)
+        {
+            for (int j = 0; j < W0->numColumns(); j++)
+            {
+                d_basis_init->item(i, j) = W0->item(i, j);
+            }
+        }
+        d_basis_init->orthogonalize();
+
+        Vector* l = NULL;
+        Vector* W0l = NULL;
+        // Find which columns of d_basis are linearly independent from W0 
+        for (int j = 0; j < d_basis->numColumns(); j++)
+        {
+            // l = W0' * u
+            Vector W_col(f_snapshots->numRows(), f_snapshots->distributed());
+            for (int i = 0; i < f_snapshots->numRows(); i++)
+            {
+                W_col.item(i) = d_basis->item(i, j);
+            }
+            l = d_basis_init->transposeMult(W_col);
+
+            // W0l = W0 * l
+            W0l = d_basis_init->mult(l);
+
+            // Compute k = sqrt(u.u - 2.0*l.l + basisl.basisl) which is ||u -
+            // basisl||_{2}.  This is the error in the projection of u into the
+            // reduced order space and subsequent lifting back to the full
+            // order space.
+            double k = W_col.inner_product(W_col) - 2.0*l->inner_product(l) +
+                      W0l->inner_product(W0l);
+            if (k <= 0)
+            {
+                k = 0;
+            }
+            else
+            {
+                k = sqrt(k);
+            }
+
+            // Use k to see if the vector addressed by u is linearly dependent
+            // on the left singular vectors.
+            if (k >= linearity_tol)
+            {
+                lin_independent_cols_W.push_back(j);
+            }
+            delete d_basis_init;
+            delete l;
```

Comment:
Instead of allocating and deleting vectors in each iteration, can work vectors be allocated before the loop? If the vectors are allocated in the multiplications, is there another version of the multiplication that uses a vector already allocated?

Reply:
They are modified now.